File: DashboardAPI/Repositories/DashboardRepository.cs
using DashboardAPI.Controllers.TransportModels;
using Npgsql;
using System.Data.SqlClient;
using System.Diagnostics;

namespace DashboardAPI.Repositories
{
    public class DashboardRepository : IDashboardRepository
    {
        private readonly NpgsqlConnection _connectionFactory;
        public DashboardRepository(NpgsqlConnection connectionFactory) 
        {
            _connectionFactory = connectionFactory;
        }

        public async Task<ReportAnalysis> GetDashboardDetails(DateTime date, int? id)
        {
            var reportAnalysis = new ReportAnalysis();
            try
            {
                await using var conn = _connectionFactory;
                await conn.OpenAsync();
                var sql = "";

                if (id != null)
                {
                    sql = $"SELECT * FROM reportanalysis WHERE id = {id}";
                }
                else
                {
                    DateTime queryDate = date;
                    sql = $"SELECT * FROM reportanalysis WHERE reportdate = '{queryDate}'";
                }

                using var cmd = new NpgsqlCommand(sql, conn);

                using var reader = cmd.ExecuteReader();

                if (!reader.Read())
                    return null;

                reportAnalysis.Date = reader.GetDateTime(1);
                reportAnalysis.TotalIncome = reader.GetDecimal(2);
                reportAnalysis.TotalExpenses = reader.GetDecimal(3);

                //then get the transactions for the report
                var reportID = reader.GetInt16(0);

                //dispose of the old reader to create another one
                reader.Close();

                var transactionSql = $"SELECT * FROM transaction WHERE report_analysis_id = {reportID}";

                await using var tranCmd = new NpgsqlCommand(transactionSql, conn);
                using var transReader = tranCmd.ExecuteReader();

                var transactionList = new List<Transaction>();



                while (transReader.Read()) 
                { 
                    var transaction = new Transaction();
                    transaction.Date = transReader.GetDateTime(2);
                    transaction.Description = transReader.GetString(3);
                    transaction.Amount = transReader.GetDecimal(4);
                    transaction.Category = transReader.GetString(5);
                    transaction.Merchant = transReader.GetString(6);
                    transaction.Month = transaction.Date.Month.ToString();
                    transactionList.Add(transaction);
                }

                reportAnalysis.CategorySummaries = CompileCategorySummary(transactionList);
                transReader.Close();
          
                return reportAnalysis;

                
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
                throw;
            }
        }

        private List<CategorySummary> CompileCategorySummary(List<Transaction> transactionList)
        {
            List<CategorySummary> categorySummary = [];
            var uniqueCategories = transactionList.Select(t => t.Category).Distinct().ToList();

            foreach (var category in uniqueCategories) 
            {
                List<Transaction> categoryTransactions = [];
                categoryTransactions = transactionList.Where(t => t.Category == category).ToList();

                var summary = new CategorySummary();
                summary.TotalAmount = categoryTransactions.Select(t => t.Amount).Sum();
                summary.Merchants = categorySummary.SelectMany(t => t.Merchants).Distinct().ToArray();
                summary.CategoryName = category;
                summary.Transactions = categoryTransactions;
                categorySummary.Add(summary);
            }

            return categorySummary;
        }

        public async Task SaveDashboardDetails(ReportAnalysis reportAnalysis)
        {
            try
            {
                await using var conn = _connectionFactory;
                
                var sw = Stopwatch.StartNew();
                await conn.OpenAsync();

                var sql = """INSERT INTO reportanalysis (report_date, total_income, total_expenses, user_id) VALUES (@date, @totalIncome, @totalExpenses, @userID) RETURNING id""";

                await using var cmd = new NpgsqlCommand(sql, conn);
                cmd.Parameters.AddWithValue("date", reportAnalysis.Date);
                cmd.Parameters.AddWithValue("totalIncome", reportAnalysis.TotalIncome);
                cmd.Parameters.AddWithValue("totalExpenses", reportAnalysis.TotalExpenses);
                //an enhancement will need to be made once there is more than one user
                cmd.Parameters.AddWithValue("userID", 1);

                var reportId = (int)await cmd.ExecuteScalarAsync();
                Console.WriteLine("Report saved to db");

                var transactions = reportAnalysis.CategorySummaries.SelectMany(cs => cs.Transactions).ToList();

                foreach (var transaction in transactions)
                {
                    //change to parameterized sql command
                    var transactionSql = """INSERT INTO transaction (report_analysis_id, date, description, amount, category, merchant, user_id) VALUES (@reportId, @transactionDate, @transactionDescription, @transactionAmount, @transactionCategory, @transactionMerchant, @userID)""";
                    await using var tranCmd = new NpgsqlCommand(transactionSql, conn);
                    tranCmd.Parameters.AddWithValue("reportId", reportId);
                    tranCmd.Parameters.AddWithValue("transactionDate", transaction.Date);
                    tranCmd.Parameters.AddWithValue("transactionDescription", transaction.Description ?? "");
                    tranCmd.Parameters.AddWithValue("transactionAmount", transaction.Amount);
                    tranCmd.Parameters.AddWithValue("transactionCategory", transaction.Category ?? "");
                    tranCmd.Parameters.AddWithValue("transactionMerchant", transaction.Merchant ?? "");
                    //an enhancement will need to be made once there is more than one user
                    tranCmd.Parameters.AddWithValue("userID", 1);
                    await tranCmd.ExecuteNonQueryAsync();
                }
                sw.Stop();
                Console.WriteLine($"Transactions saved to db in {sw.Elapsed}");

            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
                throw;
            }
        }
    }
}

File: backend/src/repositories/DashboardRepository.ts
import { PrismaClient } from '@prisma/client';
import { ReportAnalysis, CategorySummary, Transaction } from '../models/types';

export interface IDashboardRepository {
  getDashboardDetails(date: Date, id?: number | null): Promise<ReportAnalysis | null>;
  saveDashboardDetails(reportAnalysis: ReportAnalysis): Promise<void>;
}

export class DashboardRepository implements IDashboardRepository {
  private prisma: PrismaClient;

  constructor(prisma: PrismaClient) {
    this.prisma = prisma;
  }

  async getDashboardDetails(date: Date, id?: number | null): Promise<ReportAnalysis | null> {
    try {
      let report;

      if (id != null) {
        report = await this.prisma.reportAnalysis.findUnique({
          where: { id },
          include: { transactions: true }
        });
      } else {
        // Query by date
        const queryDate = new Date(date);
        queryDate.setHours(0, 0, 0, 0); // Normalize to start of day
        
        report = await this.prisma.reportAnalysis.findFirst({
          where: { 
            report_date: queryDate
          },
          include: { transactions: true }
        });
      }

      if (!report) {
        return null;
      }

      // Convert transactions to the expected format
      const transactionList: Transaction[] = report.transactions.map(t => ({
        Date: t.date,
        Description: t.description,
        Amount: Number(t.amount),
        Category: t.category,
        Merchant: t.merchant,
        Month: (t.date.getMonth() + 1).toString() // JavaScript getMonth() returns 0-11, but we need 1-12
      }));

      // Compile category summaries
      const categorySummaries = this.compileCategorySummary(transactionList);

      const reportAnalysis: ReportAnalysis = {
        Date: report.report_date,
        TotalIncome: Number(report.total_income),
        TotalExpenses: Number(report.total_expenses),
        CategorySummaries: categorySummaries
      };

      return reportAnalysis;
    } catch (error) {
      console.error('Error getting dashboard details:', error);
      throw error;
    }
  }

  private compileCategorySummary(transactionList: Transaction[]): CategorySummary[] {
    const categorySummaries: CategorySummary[] = [];
    const uniqueCategories = [...new Set(transactionList.map(t => t.Category).filter(c => c))];

    for (const category of uniqueCategories) {
      const categoryTransactions = transactionList.filter(t => t.Category === category);
      
      const merchants = [...new Set(categoryTransactions.map(t => t.Merchant).filter(m => m))];
      const totalAmount = categoryTransactions.reduce((sum, t) => sum + t.Amount, 0);

      const summary: CategorySummary = {
        CategoryName: category || '',
        Merchants: merchants as string[],
        TotalAmount: totalAmount,
        Transactions: categoryTransactions
      };

      categorySummaries.push(summary);
    }

    return categorySummaries;
  }

  async saveDashboardDetails(reportAnalysis: ReportAnalysis): Promise<void> {
    try {
      const startTime = Date.now();

      // Parse the date to ensure it's a Date object
      const reportDate = new Date(reportAnalysis.Date);
      reportDate.setHours(0, 0, 0, 0); // Normalize to start of day

      // Insert the report analysis
      const report = await this.prisma.reportAnalysis.create({
        data: {
          report_date: reportDate,
          total_income: reportAnalysis.TotalIncome,
          total_expenses: reportAnalysis.TotalExpenses
        }
      });

      console.log('Report saved to db');

      // Get all transactions from category summaries
      const transactions = reportAnalysis.CategorySummaries.flatMap(cs => cs.Transactions);

      // Insert all transactions
      for (const transaction of transactions) {
        const transactionDate = new Date(transaction.Date);
        transactionDate.setHours(0, 0, 0, 0);

        await this.prisma.transaction.create({
          data: {
            report_analysis_id: report.id,
            date: transactionDate,
            description: transaction.Description || '',
            amount: transaction.Amount,
            category: transaction.Category || '',
            merchant: transaction.Merchant || ''
          }
        });
      }

      const elapsed = Date.now() - startTime;
      console.log(`Transactions saved to db in ${elapsed}ms`);
    } catch (error) {
      console.error('Error saving dashboard details:', error);
      throw error;
    }
  }
}
